### ISSUE TITLE
Unable to grant last.fm API access

### ISSUE THREAD

**Issue by vicelversa** [ampache/ampache#1694]
# Environment

 * Ampache version: Ampache 3.8.7-develop 
 * Web server + version: Apache/2.4.29 (Unix) OpenSSL/1.0.2n PHP/7.2.3 SVN/1.9.7
 * Server operating system: x86_64-QNAP_FW43-linux-gnu
 * Client operating system: Windows 10 x64, Chrome Version 65.0.3325.146 (Official Build) (64-bit)

 * Client type:
   - Web

# Settings

```
; Last.FM API Key
; Set this to your Last.FM api key to actually use Last.FM for
; recommendations and metadata.
lastfm_api_key = "13893ba930c63b1b2cbe21441dc7f550"

; Last.FM API secret
; Set this to your Last.FM api secret to actually use Last.FM for
; scrobbling.
lastfm_api_secret = "<omitted>"
```

# Description

Unable to grant Ampache Last.fm API access.

1. Added lastfm_api_key and lastfm_api_secret to ampache.cg.php
2. enabled plugin from webgui
3. Click Last.FM Grant URL in webgui server settings - plugins
4. Approve access to connected app on last.fm
5. Ampache webgui error: Your Account has not been updated : Lastfm

# Logs

```
2018-03-14 02:58:47 [admin] (Last.FM) -> No session key, not scrobbling (need to grant Ampache to last.fm) 
2018-03-14 02:58:47 [admin] (Scrobbler) -> Cannot access https://ws.audioscrobbler.com/2.0/?method=auth.getSession&api_key=13893ba930c63b1b2cbe21441dc7f550&token=8XTDmghOx7UXWvogWpg06fD_C_yfusUI&api_sig=8439de4fca7992d66e29ca744236bb7f 
2018-03-14 02:58:47 [admin] (Last.FM) -> getSession Failed: Did not receive a valid response 
```

**Comment by wagnered** [comment]
@vicelversa 

Go to  your last.fm account and select settings from the dropdown list.

Select the "application" tab to see  the connected applications. the application name you entered in the "Create API Account" dialog should be displayed.

Perhaps, your key or secret code was entered incorrectly in ampache.cfg.php. If I remember  correctly, it took several tries to get my server connected to last.FM.

Did you get the email to verify your last.fm account and verify by clicking on the URL indicated in the message?

**Comment by vicelversa** [comment]
@wagnered 

After several more attempts (as you mentioned) it finally showed up as a connected application on my last.fm account. I was already checking there before and never saw anything.

The message still reported error each time from ampache. Perhaps there is room for improvement? Either way, not a huge issue and it appears to be working now. Thanks!

**Comment by vicelversa** [comment]
@wagnered 

I may have spoke too soon. Even though my last.fm account shows ampache as a connected application. It is still failing from the logs.

On another note, the webGUI interface keeps seeming to refresh with "loading..." in the top left every few seconds. When on the home page, I notice it seems like it's trying to load in Similar Songs as it flashes for a second then disappears. What is needed to get Similar songs (recommendations) working properly? It does seem to load similar artists.

Attached is my ampache.cfg.php and snippet of logs when playing a single song.
[ampache.cfg-CLEAN.php.txt](https://github.com/ampache/ampache/files/1812760/ampache.cfg-CLEAN.php.txt)
[ampache.20180315-SNIP.log](https://github.com/ampache/ampache/files/1812757/ampache.20180315-SNIP.log)

**Comment by wagnered** [comment]
@vicelversa 

Did you verify your email address? Are you sure that your key and secret are entered correctly in ampache.cfg.php?  If the answer to these questions is "yes" then you probably should search the [API forum](https://www.last.fm/group/Last.fm%2520Web%2520Services) or pose a question there.

I'm also running 3.8.7-develop and I have no problem scrobbling.

**Comment by vicelversa** [comment]
@wagnered 

yes verified account and yes key and secret are correct. I just created a new last.fm account to test and same issue.

I'm really not sure what the issue is here. I ended up manually entering in my last.fm-username,sessionkey in the db under the value for lastfm_grant_link. It's working.

**Comment by vicelversa** [comment]
However, the Similar Songs functionality does not seem to be working (is this actually a feature? cannot find documentation on it). I see the section load then disappear every few seconds on the home page. I will do some debugging soon on where it is breaking.

**Comment by wagnered** [comment]
> (is this actually a feature? cannot find documentation on it)

I see it references to it in the logic .No, the developers before me decided it was more important to add new features and fix bugs, than work on documentation.

 > I see the section load then disappear every few seconds 

Do you see any similar songs/artists, or just the symbol indicating it's loading something?

I still can't tell if I'm missing a setting, or there is a problem. :grimacing:

**Comment by vicelversa** [comment]
Here is a GIF of the problem. Yes similar artist load, there is a setting for this in the config file to enable / disable. However, I do not see a setting specific for similar songs. 

As seen in the GIF, as long as a song is playing, it will continue to reload the section every few seconds. Always showing similar artist but never fully loading any similar songs:

![2018-03-15_8-16-27](https://user-images.githubusercontent.com/20819767/37472554-9ae74a8e-2829-11e8-872b-e007dd884d4b.gif)

A still shot of the quick "loading" moment:

![2018-03-15_8-18-47](https://user-images.githubusercontent.com/20819767/37472597-b31ada4e-2829-11e8-98cc-5279974f72e2.png)

**Comment by wagnered** [comment]
@vicelversa 

There is also a `refresh_limit` switch in the config file. It defaults to 60 seconds, though.

Are you using "localplay"?

I found one of my problems: javaScript code only allows desktop notification if logged in to loclhost.

There is still another problem I have to track down.

**Comment by vicelversa** [comment]
@wagnered 

Not using localplay, using the Web Player.

The refresh_limit is configured with the default value (60)

**Comment by wagnered** [comment]
@vicelversa 

Success: Was able to replicate your issue.  And looks like last.FM isn't finding similar songs. 

 If the song has an mbid the query uses that: If not, it uses the track name.

I noticed that in the similar artist list, Ampache will highlight artists that are in the catalog(s).  Cool! :sunglasses: 

So it looks like it's working as designed...

**Comment by vicelversa** [comment]
@wagnered 

Thank you for testing. In my example song above in the screenshot. I certainly have similar songs in my library which match that track when manually searching on last.fm

However, I do not have any mbid information in the tracks in my library. Unfortunately I cannot simply add them via Ampache as that function seems to be broken at the moment. I see several issues open on it, not sure yet what aspect is affecting my instance.

Will the similar songs function work only if there is an exact match on last.fm?
Basically, does this function (similar songs) work for you in any scenario from your testing?

**Comment by wagnered** [comment]
@vicelversa 

> Unfortunately I cannot simply add them via Ampache as that function seems to be broken at the moment. 

Never was a feature of Ampache :smile: 

> Will the similar songs function work only if there is an exact match on last.fm?

I don't know much about Last.FM.  but my tracks on Last.FM show Benjamin Booker's "Violent Shiver" as similar to Alabama Shakes, but it doesn't indicate what Alabama Shake song it is. And Benjamin Booker is not on Ampache's  list of Similar artists. This is the most time i've ever spent dealing with Last.FM.   I only added that account so I could debug issues.

I don't think Last.FM is very in matching similar song/artists.

You could try Musicbrianz Picard to add the mbid's to a test set of files  and  see if Last.FM is better at finding similar songs that way.

I stepped through the code so I know that a query is being sent to last.FM and returns 0 results.

**Comment by wagnered** [comment]
@vicelversa 

I just added musicbrainz id's to about 20 artists and their albums.  Ampache is now displaying Similar songs.

Do you require a session? Are `require_session` and/or `require_localnet_session` enabled and set to true in ampache.cfg.php?  It appears that a session key is required.

**Comment by mitchray** [comment]
Closing due to inactivity, reopen if this still applies to the latest version
